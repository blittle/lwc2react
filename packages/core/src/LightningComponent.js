import React from 'react';
import ObservableMembrane from 'observable-membrane';

export default class LightningComponent extends React.Component {
  constructor(template) {
    super();

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted) this.forceUpdate();
      },
    });

    if (!this.__s) {
      this.__s = membrane.getProxy(Object.assign({}, this.props));
    }

    this.template = React.createRef();

    template.customEvents.forEach((event) => {
      const ref = event[2];
      this[ref] = React.createRef();
    });
  }

  componentDidUpdate(prevProps) {
    this.updateMembrane(prevProps);
  }

  componentDidMount(template) {
    this.mounted = true;
    this.updateMembrane();
    this.addLocalStyles(template);
    this.addEventHandlers(template);
  }

  componentWillUnmount(template) {
    this.mounted = false;
    this.cleanupLocalStyles(template);
    this.cleanupEventHandlers(template);
  }

  addLocalStyles(template) {
    this.stylesheets = [];
    template.stylesheets.forEach((stylesheet) => {
      const sheet = document.createElement('style');
      sheet.type = 'text/css';
      sheet.textContent = stylesheet(
        '[' + template.stylesheetTokens.hostAttribute.toLowerCase() + ']',
        '[' + template.stylesheetTokens.shadowAttribute.toLowerCase() + ']',
        null
      );
      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });
  }

  addEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[2];
      this[cb] = this[cb].bind(this);
      if (ref !== 'template') {
        this[ref].current.template.current.addEventListener(name, this[cb]);
      } else {
        this[ref].current.addEventListener(name, this[cb]);
      }
    });
  }

  cleanupLocalStyles() {
    this.stylesheets.forEach((sheet) => {
      if (sheet.parentNode) sheet.parentNode.removeChild(sheet);
    });
  }

  cleanupEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[3];

      if (ref !== 'template') {
        this[ref].current.template.current.removeEventListener(name, this[cb]);
      } else {
        this[ref].current.removeEventListener(name, this[cb]);
      }
    });
  }

  updateMembrane(prevProps = {}) {
    Object.keys(this.props).forEach((prop) => {
      if (prevProps[prop] !== this.props[prop])
        this.__s[prop] = this.props[prop];
    });
  }
}
