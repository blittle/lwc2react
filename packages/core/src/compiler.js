import { parse } from 'recast';
import { compileClass } from './class-compiler';
import { compileTemplate } from './template-compiler';

export function compile(id, source) {
  if (id.includes('@lwc/engine/dist/engine.js') || id.includes('wire-service'))
    return 'export default undefined';

  const ast = parse(source);

  if (id.endsWith('.html')) {
    return compileTemplate(ast);
  } else if (id.endsWith('.css')) {
    return source;
  } else {
    return compileClass(ast);
  }
}
