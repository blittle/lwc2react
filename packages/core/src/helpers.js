import { parse, types } from 'recast';

// const b = types.builders;
const n = types.namedTypes;

export function removeLWCCode(ast) {
  const lwcIndex1 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );
  ast.program.body[lwcIndex1] = parse(
    'import React from "react"'
  ).program.body[0];

  let lwcIndex2 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );

  while (lwcIndex2 !== -1) {
    ast.program.body.splice(lwcIndex2, 1);
    lwcIndex2 = ast.program.body.findIndex(
      (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
    );
  }

  const registerDecoratorsIndex = ast.program.body.findIndex(
    (el) =>
      n.ExpressionStatement.check(el) &&
      n.CallExpression.check(el.expression) &&
      el.expression.callee.name === '_registerDecorators'
  );

  if (registerDecoratorsIndex !== -1) {
    ast.program.body.splice(registerDecoratorsIndex, 1);
  }
}
