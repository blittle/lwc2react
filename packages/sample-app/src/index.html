<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title></title>
    <link rel="stylesheet" href="/css/global.css" />
    <link rel="stylesheet" href="https://blittle-developer-edition.ap1.stmpb.stm.force.com/assets/styles/salesforce-lightning-design-system.min.css?25def0b3bc" />
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="dist/index.js"></script>
  </body>
</html>
