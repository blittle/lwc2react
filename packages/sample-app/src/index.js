import MyApp from 'my/home';
import React from 'react';
import ReactDOM from 'react-dom';

ReactDOM.render(React.createElement(MyApp), document.getElementById('app'));
