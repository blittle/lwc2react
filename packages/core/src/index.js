import pluginUtils from '@rollup/pluginutils';
import { compile } from './compiler';

export default function rollupLWC2ReactCompiler(pluginOptions = {}) {
  let { include, exclude, debug } = pluginOptions;

  const filter = pluginUtils.createFilter(include, exclude);

  return {
    name: 'rollup-lwc2react-compiler',
    async transform(src, id) {
      if (debug && filter(id)) {
        console.log('lwc2react -> ', id);
      }

      if (!filter(id)) {
        return;
      }

      let code;
      try {
        code = compile(id, src);
        if (code) {
          if (debug && id.includes(debug)) {
            console.log(src);
            console.log('===========');
            console.log(code);
          }
          return {
            code,
            map: null,
          };
        }
      } catch (error) {
        console.log('error processing lwc source: ', id);
        console.log(src);
        throw error;
      }
    },
  };
}
