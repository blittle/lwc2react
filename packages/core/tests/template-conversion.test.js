import { compile } from '../src/compiler';

describe('templates', function () {
  it('should compile an empty component', function () {
    const source = `
import _implicitStylesheets from "./test.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {} = $api;
  return [];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile div with text', function () {
    const source = `
import _implicitStylesheets from "./test.css";
import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [api_element("div", {
    key: 0
  }, [api_text("hi")])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile multiple nested divs with text', function () {
    const source = `
import _implicitStylesheets from "./test.css";
import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [api_element("div", {
    key: 2
  }, [api_element("div", {
    key: 0
  }, [api_text("hi 1")]), api_element("div", {
    key: 1
  }, [api_text("hi 2")])])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile class attributes', function () {
    const source = `
import _implicitStylesheets from "./test.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [api_element("div", {
    classMap: {
      "hello": true
    },
    key: 0
  }, [api_text("text")])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile style attributes', function () {
    const source = `
import _implicitStylesheets from "./test.css";
import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [api_element("div", {
    styleMap: {
      "color": "red"
    },
    key: 0
  }, [api_text("text")])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile generic attributes', function () {
    const source = `
import _implicitStylesheets from "./test.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [api_element("div", {
    attrs: {
      "title": "wow"
    },
    key: 0
  }, [api_text("text")])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile complex heirarchy with attributes', function () {
    const code = `
import _implicitStylesheets from "./test.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    h: api_element,
    t: api_text,
    gid: api_scoped_id
  } = $api;
  return [api_element("figure", {
    attrs: {
      "role": "group"
    },
    key: 3
  }, [api_element("img", {
    styleMap: {
      "border": "1px",
      "borderColor": "red"
    },
    attrs: {
      "src": "operahousesteps.jpg",
      "alt": "The Sydney Opera House"
    },
    key: 0
  }, []), api_element("figcaption", {
    classMap: {
      "something": true,
      "else": true
    },
    key: 2
  }, [api_text("We saw the opera "), api_element("cite", {
    key: 1
  }, [api_text("Barber of Seville")]), api_text(" here!")])])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-test_test-host",
  shadowAttribute: "my-test_test"
};
        `.trim();

    expect(compile('something.html', code)).toMatchSnapshot();
  });

  it('should compile the lwc class into a react class that renders the template', function () {
    const source = `
import _tmpl from "./test.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';

class Test extends LightningElement {}

export default _registerComponent(Test, {
  tmpl: _tmpl
});
    `;

    expect(compile('something.js', source)).toMatchSnapshot();
  });

  it('should generate the proper template with api bound props', function () {
    const source = `
import _implicitStylesheets from "./product.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    d: api_dynamic,
    h: api_element
  } = $api;
  return [api_element("h1", {
    key: 0
  }, [api_text("Product: "), api_dynamic($cmp.params.productId)])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-product_product-host",
  shadowAttribute: "my-product_product"
};
`;
    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile conditional blocks and expressions', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./app.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';

class ProductImage extends LightningElement {
  constructor(...args) {
    super(...args);
    this.someValue = 1;
  }

  change() {
    const someValue = this.someValue;

    if (this.someValue) {
      this.someValue = this.someValue + 1;
    } else {
      this.someValue = someValue + 1;
    }

    this.someValue = this.someValue ? this.someValue + 1 : someValue + 1;
  }

}

_registerDecorators(ProductImage, {
  fields: ["someValue"]
})

export default _registerComponent(ProductImage, {
  tmpl: _tmpl
});
    `;

    expect(compile('some.js', source)).toMatchSnapshot();
  });

  it('should compile for loops', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./app.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';

class ProductImage extends LightningElement {
  constructor(...args) {
    super(...args);
    this.someValue = 1;
  }

  change() {
    for (let i = this.someValue; i < this.someValue + 1; this.someValue++) {
      this.someValue = this.someValue + 1;
    }
  }

}

_registerDecorators(ProductImage, {
  fields: ["someValue"]
})

export default _registerComponent(ProductImage, {
  tmpl: _tmpl
});
    `;

    expect(compile('some.js', source)).toMatchSnapshot();
  });

  it('should compile forof loops', function () {
    const source = `
import { registerDecorators as _registerDecorators } from "lwc";
import _tmpl from "./app.html";
import { registerComponent as _registerComponent } from "lwc";
import { LightningElement } from 'lwc';

class ProductImage extends LightningElement {
  constructor(...args) {
    super(...args);
    this.arr = [1, 2];
  }

  change() {
    for (let el of this.arr) {
      this.arr.push(this.something);
    }
  }

}

_registerDecorators(ProductImage, {
  fields: ["someValue"]
})

export default _registerComponent(ProductImage, {
  tmpl: _tmpl
});
    `;

    expect(compile('some.js', source)).toMatchSnapshot();
  });

  it('should compile for:each directives', function () {
    const source = `
import _implicitStylesheets from "./app.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    d: api_dynamic,
    k: api_key,
    h: api_element,
    i: api_iterator
  } = $api;
  return [api_element("ul", {
    key: 1
  }, api_iterator($cmp.list, function (item) {
    return api_element("li", {
      key: api_key(0, item.id)
    }, [api_dynamic(item.value)]);
  }))];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-app_app-host",
  shadowAttribute: "my-app_app"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile for:each directives with dynamic attributes', function () {
    const source = `
import _implicitStylesheets from "./app.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    d: api_dynamic,
    h: api_element,
    k: api_key,
    i: api_iterator
  } = $api;
  return [api_element("ul", {
    key: 2
  }, api_iterator($cmp.list, function (item) {
    return api_element("li", {
      className: item.value,
      attrs: {
        "title": item.value
      },
      key: api_key(1, item.id)
    }, [api_element("span", {
      className: $cmp.classProp,
      key: 0
    }, [api_dynamic(item.value)])]);
  }))];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-app_app-host",
  shadowAttribute: "my-app_app"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });

  it('should compile if:true directives', function () {
    const source = `
import _implicitStylesheets from "./app.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    h: api_element
  } = $api;
  return [$cmp.someProp ? api_element("div", {
    key: 0
  }, [api_text("a")]) : null, !$cmp.someProp ? api_element("div", {
    key: 1
  }, [api_text("b")]) : null];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-app_app-host",
  shadowAttribute: "my-app_app"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });
  it('should convert ID attributes', function () {
    const source = `
import _implicitStylesheets from "./app.css";

import { registerTemplate } from "lwc";

function tmpl($api, $cmp, $slotset, $ctx) {
  const {
    t: api_text,
    gid: api_scoped_id,
    h: api_element
  } = $api;
  return [api_element("div", {
    attrs: {
      "id": api_scoped_id("1")
    },
    key: 0
  }, [api_text("content")])];
}

export default registerTemplate(tmpl);
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets)
}
tmpl.stylesheetTokens = {
  hostAttribute: "my-app_app-host",
  shadowAttribute: "my-app_app"
};
        `.trim();

    expect(compile('something.html', source)).toMatchSnapshot();
  });
});
