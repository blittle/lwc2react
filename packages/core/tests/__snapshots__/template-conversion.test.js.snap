// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates should compile an empty component 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return null;
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile class attributes 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template,
    className: \\"hello\\"
  }, \\"text\\");
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile complex heirarchy with attributes 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"figure\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 3,
    ref: $cmp.template,
    \\"role\\": \\"group\\"
  }, [React.createElement(\\"img\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    \\"src\\": \\"operahousesteps.jpg\\",
    \\"alt\\": \\"The Sydney Opera House\\",

    style: {
      \\"border\\": \\"1px\\",
      \\"borderColor\\": \\"red\\"
    }
  }, null), React.createElement(\\"figcaption\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 2,
    className: \\"something else\\"
  }, [\\"We saw the opera \\", React.createElement(\\"cite\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1
  }, \\"Barber of Seville\\"), \\" here!\\"])]);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile conditional blocks and expressions 1`] = `
"import LightningComponent from \\"lwc2react/lib/LightningComponent\\";
import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./app.html\\";

class ProductImage extends LightningComponent {
  constructor(...args) {
    super(_tmpl);
    this.__s.someValue = 1;
  }

  change() {
    const someValue = this.__s.someValue;

    if (this.__s.someValue) {
      this.__s.someValue = this.__s.someValue + 1;
    } else {
      this.__s.someValue = someValue + 1;
    }

    this.__s.someValue = this.__s.someValue ? this.__s.someValue + 1 : someValue + 1;
  }

  render() {
    const prox = Object.assign({}, this, this.__s);
    prox.__proto__ = this.__proto__;
    return _tmpl(prox);
  }

  componentDidMount() {
    super.componentDidMount(_tmpl);
  }

  componentWillUnmount() {
    super.componentWillUnmount(_tmpl);
  }
}

export default ProductImage;"
`;

exports[`templates should compile div with text 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template
  }, \\"hi\\");
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile for loops 1`] = `
"import LightningComponent from \\"lwc2react/lib/LightningComponent\\";
import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./app.html\\";

class ProductImage extends LightningComponent {
  constructor(...args) {
    super(_tmpl);
    this.__s.someValue = 1;
  }

  change() {
    for (let i = this.__s.someValue; i < this.__s.someValue + 1; this.__s.someValue++) {
      this.__s.someValue = this.__s.someValue + 1;
    }
  }

  render() {
    const prox = Object.assign({}, this, this.__s);
    prox.__proto__ = this.__proto__;
    return _tmpl(prox);
  }

  componentDidMount() {
    super.componentDidMount(_tmpl);
  }

  componentWillUnmount() {
    super.componentWillUnmount(_tmpl);
  }
}

export default ProductImage;"
`;

exports[`templates should compile for:each directives 1`] = `
"import _implicitStylesheets from \\"./app.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"ul\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1,
    ref: $cmp.template
  }, $cmp.list.map(item => React.createElement(\\"li\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: item.id
  }, item.value)));
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-app_app-host\\",
  shadowAttribute: \\"my-app_app\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile for:each directives with dynamic attributes 1`] = `
"import _implicitStylesheets from \\"./app.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"ul\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 2,
    ref: $cmp.template
  }, $cmp.list.map(item => React.createElement(\\"li\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: item.id,
    \\"title\\": item.value,
    className: item.value
  }, React.createElement(\\"span\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    className: $cmp.classProp
  }, item.value))));
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-app_app-host\\",
  shadowAttribute: \\"my-app_app\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile forof loops 1`] = `
"import LightningComponent from \\"lwc2react/lib/LightningComponent\\";
import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./app.html\\";

class ProductImage extends LightningComponent {
  constructor(...args) {
    super(_tmpl);
    this.__s.arr = [1, 2];
  }

  change() {
    for (let el of this.__s.arr) {
      this.__s.arr.push(this.__s.something);
    }
  }

  render() {
    const prox = Object.assign({}, this, this.__s);
    prox.__proto__ = this.__proto__;
    return _tmpl(prox);
  }

  componentDidMount() {
    super.componentDidMount(_tmpl);
  }

  componentWillUnmount() {
    super.componentWillUnmount(_tmpl);
  }
}

export default ProductImage;"
`;

exports[`templates should compile generic attributes 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template,
    \\"title\\": \\"wow\\"
  }, \\"text\\");
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile if:true directives 1`] = `
"import _implicitStylesheets from \\"./app.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return [$cmp.someProp ? React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0
  }, \\"a\\") : null, !$cmp.someProp ? React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1
  }, \\"b\\") : null];
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-app_app-host\\",
  shadowAttribute: \\"my-app_app\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile multiple nested divs with text 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 2,
    ref: $cmp.template
  }, [React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0
  }, \\"hi 1\\"), React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 1
  }, \\"hi 2\\")]);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile style attributes 1`] = `
"import _implicitStylesheets from \\"./test.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template,

    style: {
      \\"color\\": \\"red\\"
    }
  }, \\"text\\");
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-test_test-host\\",
  shadowAttribute: \\"my-test_test\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should compile the lwc class into a react class that renders the template 1`] = `
"import LightningComponent from \\"lwc2react/lib/LightningComponent\\";
import ObservableMembrane from \\"observable-membrane\\";
import _tmpl from \\"./test.html\\";
import React from \\"react\\";

class Test extends LightningComponent {
  constructor() {
    super(_tmpl);
  }

  render() {
    const prox = Object.assign({}, this, this.__s);
    prox.__proto__ = this.__proto__;
    return _tmpl(prox);
  }

  componentDidMount() {
    super.componentDidMount(_tmpl);
  }

  componentWillUnmount() {
    super.componentWillUnmount(_tmpl);
  }
}

export default Test;"
`;

exports[`templates should convert ID attributes 1`] = `
"import _implicitStylesheets from \\"./app.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"div\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template,
    \\"id\\": \\"1\\"
  }, \\"content\\");
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-app_app-host\\",
  shadowAttribute: \\"my-app_app\\"
};

tmpl.customEvents = [];"
`;

exports[`templates should generate the proper template with api bound props 1`] = `
"import _implicitStylesheets from \\"./product.css\\";
import React from \\"react\\";

function tmpl($cmp) {
  return React.createElement(\\"h1\\", {
    [tmpl.stylesheetTokens.shadowAttribute.toLowerCase()]: \\"true\\",
    key: 0,
    ref: $cmp.template
  }, [\\"Product: \\", $cmp.params.productId]);
}

export default tmpl;
tmpl.stylesheets = [];

if (_implicitStylesheets) {
  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);
}

tmpl.stylesheetTokens = {
  hostAttribute: \\"my-product_product-host\\",
  shadowAttribute: \\"my-product_product\\"
};

tmpl.customEvents = [];"
`;
